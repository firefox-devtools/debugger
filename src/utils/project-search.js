/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

// Maybe reuse file search's functions?

import React from "react";
import type { Match } from "../components/ProjectSearch";

export function highlightMatches(lineMatch: Match) {
  const { value, column, match } = lineMatch;
  const len = match.length;

  return (
    <span className="line-value">
      <span className="line-match" key={0}>
        {value.slice(0, column)}
      </span>
      <span className="query-match" key={1}>
        {value.substr(column, len)}
      </span>
      <span className="line-match" key={2}>
        {value.slice(column + len, value.length)}
      </span>
    </span>
  );
}
