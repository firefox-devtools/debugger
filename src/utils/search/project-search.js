// Maybe reuse file search's functions?
export function findSourceMatches(source, queryText) {
  const { text, loading } = source;
  if (loading || !text || queryText == "") {
    return [];
  }

  const lines = text.split("\n");
  let result = undefined;
  const query = new RegExp(queryText, "g");

  let matches = lines
    .map((_text, line) => {
      const indices = [];

      while ((result = query.exec(_text))) {
        indices.push({
          sourceId: source.id,
          line: line + 1,
          column: result.index,
          match: result[0],
          value: _text,
          text: result.input
        });
      }
      return indices;
    })
    .filter(_matches => _matches.length > 0);

  matches = [].concat(...matches);
  return matches;
}
