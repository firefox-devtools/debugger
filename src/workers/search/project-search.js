// Maybe reuse file search's functions?
import { isLoaded } from "../../utils/source";

export function findSourceMatches(source, queryText) {
  const { text } = source;
  if (!isLoaded(source) || !text || queryText == "") {
    return [];
  }

  const lines = text.split("\n");
  let result = undefined;
  const query = new RegExp(queryText, "g");

  let matches = lines
    .map((_text, line) => {
      const indices = [];

      while ((result = query.exec(_text))) {
        indices.push({
          sourceId: source.id,
          line: line + 1,
          column: result.index,
          match: result[0],
          value: _text,
          text: result.input
        });
      }
      return indices;
    })
    .filter(_matches => _matches.length > 0);

  matches = [].concat(...matches);
  return matches;
}
