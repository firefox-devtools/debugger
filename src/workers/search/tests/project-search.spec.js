/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import { findSourceMatches } from "../project-search";

const text = `
  function foo() {
    foo();
  }
`;

describe("project search", () => {
  it("simple", () => {
    const source = {
      text,
      loadedState: "loaded",
      id: "bar.js",
      url: "http://example.com/foo/bar.js"
    };

    const matches = findSourceMatches(source, "foo");
    expect(matches).toMatchSnapshot();
  });
});
