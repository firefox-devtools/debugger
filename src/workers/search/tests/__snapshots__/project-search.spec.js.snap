// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`project search finds matches 1`] = `
Array [
  Object {
    "column": 11,
    "line": 2,
    "match": "foo",
    "sourceId": "bar.js",
    "value": "  function foo() {",
  },
  Object {
    "column": 4,
    "line": 3,
    "match": "foo",
    "sourceId": "bar.js",
    "value": "    foo();",
  },
]
`;
